### ISSUE TITLE
rustc 1.87.0 ICE compiling rusqlite 0.31.0 in Docker release build

### ISSUE THREAD

**Issue by coe0718** [rust-lang/rust#155369]
### Summary

I hit an internal compiler error from `rustc 1.87.0` while building `rusqlite 0.31.0` inside a Docker release build.

The crash happened repeatedly while compiling `rusqlite` with `feature="bundled"` and involved repeated panics in:

- `compiler/rustc_trait_selection/src/traits/specialize/specialization_graph.rs:41:57`
- `compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs:563:83`

The panic message was:

```text
called `Option::unwrap()` on a `None` value
```

This was reported by `rustc` itself as:

```text
error: the compiler unexpectedly panicked. this is a bug.
```

As of April 15, 2026, I can no longer reproduce this reliably from the same checkout with the same `rust:1.87-slim` tag, so this may be an unstable or dependency-resolution-sensitive ICE. I am filing it anyway because the original failure log is very detailed and points at compiler internals.

### Reproducer

The failing build came from this Dockerfile:

```Dockerfile
FROM rust:1.87-slim AS builder
WORKDIR /app
RUN apt-get update && apt-get install -y pkg-config libssl-dev && rm -rf /var/lib/apt/lists/*
COPY Cargo.toml .
COPY src ./src
RUN cargo build --release
```

The crate under build depended on:

```toml
rusqlite = { version = "0.31", features = ["bundled"] }
```

The failure occurred during:

```text
RUN cargo build --release
```

and the final error was:

```text
error: could not compile `rusqlite` (lib)
```

One of the rustc invocations from the failing log was:

```text
/usr/local/rustup/toolchains/1.87.0-x86_64-unknown-linux-gnu/bin/rustc --crate-name rusqlite --edition=2021 /usr/local/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/rusqlite-0.31.0/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -C embed-bitcode=no --cfg 'feature="bundled"' --cfg 'feature="modern_sqlite"' ... -C strip=debuginfo ...
```

Relevant resolved crates from the failure log included:

- `rusqlite 0.31.0`
- `bitflags 2.11.1`
- `libsqlite3-sys 0.28.0`
- `hashlink 0.9.1`
- `smallvec 1.15.1`

### What I expected

`cargo build --release` should either compile successfully or produce a normal diagnostic from the crate graph. It should not ICE in trait selection.

### What happened instead

`rustc` repeatedly panicked while compiling `rusqlite`, including query stacks like:

```text
#0 [specialization_graph_of] building specialization graph of trait `core::ops::drop::Drop`
#1 [coherent_trait] coherence checking all impls of trait `core::ops::drop::Drop`
```

and:

```text
#0 [evaluate_obligation] evaluating trait selection obligation `<i32 as core::ops::bit::BitXor>::Output == i32`
#1 [compare_impl_item] checking assoc item `<impl at /usr/local/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/bitflags-2.11.1/src/public.rs:536:9: 536:47>::Bits` is compatible with trait definition
```

The repeated panic sites were:

```text
thread 'rustc' panicked at compiler/rustc_trait_selection/src/traits/specialize/specialization_graph.rs:41:57:
called `Option::unwrap()` on a `None` value
```

and later:

```text
thread 'rustc' panicked at compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs:563:83:
called `Option::unwrap()` on a `None` value
```

I saw many repeated query stacks for traits like:

- `core::ops::drop::Drop`
- `core::marker::Copy`
- `core::clone::Clone`
- `core::fmt::Debug`
- `core::marker::Send`
- `core::cmp::Eq`
- `core::hash::Hash`
- `core::cmp::PartialEq`
- `core::cmp::PartialOrd`
- `core::cmp::Ord`
- `core::ops::bit::BitXor`
- `core::ops::bit::BitXorAssign`
- `core::ops::bit::BitAnd`
- `core::ops::bit::BitAndAssign`
- `core::ops::arith::Sub`
- `core::ops::arith::SubAssign`
- `core::ops::bit::Not`
- `core::marker::Sync`
- `core::error::Error`
- `core::ops::deref::Deref`
- `core::ops::deref::DerefMut`
- `core::convert::TryFrom`

### Version

From the failing output:

```text
rustc 1.87.0 (17067e9ac 2025-05-09) running on x86_64-unknown-linux-gnu
```

I also confirmed on April 15, 2026 that `rust:1.87-slim` currently reports:

```text
rustc 1.87.0 (17067e9ac 2025-05-09)
host: x86_64-unknown-linux-gnu
LLVM version: 20.1.1
```

### Additional context

- The build was happening in Docker, not in a local interactive shell.
- The builder stage used `rust:1.87-slim`.
- The original failure log was very large because the compiler panicked repeatedly across many trait queries before the build stopped.
- On April 15, 2026, `docker build -q -f products/signal-hive/backend/Dockerfile products/signal-hive/backend` succeeded for me from the same checkout, so I do not currently have a stable minimized reproducer.

**Comment by apiraino** [comment]
Quickly tried building a new crate pulling rusqlite as dependency and compiled with `cargo +1.87.0 build --release`, just to exclude some weird conditions that would make that crate not build anymore. The crate builds so it must be something else.

So @coe0718 this needs a reduction (possibly removing Docker from the equation and using just cargo). If you have a chance to figure out a reproducible, that would be appreciated. Thanks!

@rustbot label E-needs-mcve

**Comment by coe0718** [comment]
I tried a build again, didn’t change any code and it compiled fine. I haven’t been able to replicate it again.
